core/src/apps/bitcoin/sign_tx/tx_info.py

```diff
@@ -0,0 +1,178 @@
+_SEQUENCE_FINAL = const(0xFFFFFFFF)
+
+
+class TxInfoBase:
+    def __init__(self, signer: Signer) -> None:
+        # Checksum of multisig inputs, used to validate change-output.
+        self.multisig_fingerprint = MultisigFingerprintChecker()
+
+        # Common prefix of input paths, used to validate change-output.
+        self.wallet_path = WalletPathChecker()
+
+        # h_tx_check is used to make sure that the inputs and outputs streamed in
+        # different steps are the same every time, e.g. the ones streamed for approval
+        # in Steps 1 and 2 and the ones streamed for signing legacy inputs in Step 4.
+        self.h_tx_check = HashWriter(sha256())  # not a real tx hash
+
+        # BIP-0143 transaction hashing.
+        self.hash143 = signer.create_hash143()
+
+        # The minimum nSequence of all inputs.
+        self.min_sequence = _SEQUENCE_FINAL
+
+    def add_input(self, txi: TxInput) -> None:
+        self.hash143.add_input(txi)  # all inputs are included (non-segwit as well)
+        writers.write_tx_input_check(self.h_tx_check, txi)
+        self.min_sequence = min(self.min_sequence, txi.sequence)
+
+        if not input_is_external(txi):
+            self.wallet_path.add_input(txi)
+            self.multisig_fingerprint.add_input(txi)
+
+    def add_output(self, txo: TxOutput, script_pubkey: bytes) -> None:
+        self.hash143.add_output(txo, script_pubkey)
+        writers.write_tx_output(self.h_tx_check, txo, script_pubkey)
+
+    def check_input(self, txi: TxInput) -> None:
+        self.wallet_path.check_input(txi)
+        self.multisig_fingerprint.check_input(txi)
+
+    def output_is_change(self, txo: TxOutput) -> bool:
+        if txo.script_type not in common.CHANGE_OUTPUT_SCRIPT_TYPES:
+            return False
+        if txo.multisig and not self.multisig_fingerprint.output_matches(txo):
+            return False
+        return (
+            self.wallet_path.output_matches(txo)
+            and len(txo.address_n) >= BIP32_WALLET_DEPTH
+            and txo.address_n[-2] <= _BIP32_CHANGE_CHAIN
+            and txo.address_n[-1] <= _BIP32_MAX_LAST_ELEMENT
+            and txo.amount > 0
+        )
+
+    def lock_time_disabled(self) -> bool:
+        return self.min_sequence == _SEQUENCE_FINAL
+
+    def get_tx_check_digest(self) -> bytes:
+        return self.h_tx_check.get_digest()
+
+
+# Used to keep track of the transaction currently being signed.
+class TxInfo(TxInfoBase):
+    def __init__(self, signer: Signer, tx: SignTx) -> None:
+        super().__init__(signer)
+        self.tx = tx
+
+
+# Used to keep track of any original transactions which are being replaced by the current transaction.
+class OriginalTxInfo(TxInfoBase):
+    def __init__(self, signer: Signer, tx: PrevTx, orig_hash: bytes) -> None:
+        super().__init__(signer)
+        self.tx = tx
+        self.signer = signer
+        self.orig_hash = orig_hash
+
+        # The current index when iterating over inputs or outputs.
```

Comment:
Please expand this comment. Use of `index` is really unclear before you fully understand the replacement signing flow.

Something like:
> Index of the next input or output to be added by `add_input` or `add_output`. Signer uses this value to check that original transaction inputs and outputs are streamed in order, and to check whether any have been skipped.

perhaps also something like "`index` is technically a variable that belongs to the signer class, and is attached to this class for convenience". But that seems to help _my_ understanding _now_, and i'm not sure if anything like this would help a future reader......

Also there should be a comment somewhere around here that explains how the class is used, i.e., that first all inputs are fed to `add_input`, then all outputs to `add_output`, and that it doesn't make sense to mess up the order.

Reply:
fwiw it would be nicer for `index` to be handled privately in `add_input/output`, but that seems impractical (we'd need more arguments and more methods)

Reply:
https://github.com/trezor/trezor-firmware/pull/1292/commits/9167c044c2f64636093efca068ac1f69ce34fbf9